[CodeGen] Expose the extensibility of PassConfig to plugins

Tcc100 opened the pull request:
This PR exposes the backend pass config to plugins via a callback. Plugin authors can register a callback that is being triggered before the target backend adds their passes to the pipeline. In the callback they then get access to the `TargetMachine`, the `PassManager`, and the `TargetPassConfig`. This allows plugins to call `TargetPassConfig::insertPass`, which is honored in the subsequent `addPass` of the main backend. We implemented this using the legacy pass manager as the backend is still using the old pass manager.

The following example shows how plugin authors can use the callback. Since its a callback that is not doing anything without anybody registering it, there shouldn't be any potential harm to the compiler unless a plugin is present.

```cpp
__attribute__((constructor)) static void initCodeGenPlugin() {
    initializeCodeGenTestPass(*PassRegistry::getPassRegistry());

    TargetMachine::registerTargetPassConfigCallback([](auto &TM, auto &PM, auto *TPC) {
        TPC->insertPass(&GCLoweringID, &CodeGenTest::ID);
    });
}
```

antoniofrighetto reviewed (commented):
Missing a test for this?

arsenm commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -72,6 +74,10 @@ namespace yaml {
 
+class TargetMachine;
+using PassConfigCallback =
+    std::function<void(TargetMachine &, PassManagerBase &, TargetPassConfig *)>;
```
function_ref?

arsenm commented on llvm/lib/CodeGen/CodeGenTargetMachineImpl.cpp:
```diff
@@ -119,6 +119,9 @@ addPassesToGenerateCode(CodeGenTargetMachineImpl &TM, PassManagerBase &PM,
   PM.add(PassConfig);
   PM.add(&MMIWP);
 
+  for (auto& C : *TargetMachine::TargetPassConfigCallbacks)
```
no auto (and no &?)

Tcc100 replied:
This is same syntax as most of the callbacks in the code base use. Do you think not using auto would be better in this case?

antoniofrighetto replied:
I think it's fine to use auto when code readability benefits from not spelling it out, which should be the case for callable objects (the container TargetPassConfigCallbacks already hints it is a callback item). That said, since we have an alias for the std::function here, `PassConfigCallback &C` should be fine too. @Tcc100 Could you also please run clang-format here? This should have been `auto &C`.

weliveindetail replied:
Agree. But it's not "almost never", but https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

Tcc100 replied:
Ok, I used `PassConfigCallback &C` as @antoniofrighetto suggested!

arsenm reviewed (commented):
insertPass is a bad API (and probably going away in the new PM?). Should probably have a set of insert points like the new pass manager does for middle end passes. Also, needs tests / a sample and documentation

Tcc100 commented:
I added a test case and some docs, which tuned out to be more challenging than expected, since I couldn't find any test case for the plugin system. The lit test in this PR requires the current llvm headers, which are not provided by the lit system, thus, it currently only works if you provide the absolute path in your system. I'd appreciate any pointers on how to implement this correctly.

arsenm commented:
I don't think you can reasonably test this with a lit test, you have to add this as a proper build target in llvm/examples similar to Bye

arsenm commented:
> I don't think you can reasonably test this with a lit test, you have to add this as a proper build target in llvm/examples similar to Bye.

A lit test could then drive the load of that plugin, but the lit test can't contain the pass itself

weliveindetail reviewed (commented):
Hi all, I am no expert for the legacy pass manager, so let's wait for the others to add their thoughts. Overall this patch LGTM.

While most folks focus on porting MachineIR passes to the NewPM these days, this patch still seems like a reasonable addition. IIUC the legacy PM is still the default infra for codegen and will likely stay around for a while.

Unfortunately, we don't have good end-to-end testing for plugins in general. The proposed unittests repeat what we have in other places. Good enough. Could we use them for drafting similar functionality in the NewPM in a follow-up step? Might be good timing since the NewPM work is in progress right now.

weliveindetail commented:
@arsenm Thanks for your follow-up. It's mostly style details, but what about the general direction of the patch?

weliveindetail commented on llvm/unittests/CodeGen/CGPluginTest/PluginTest.cpp:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  TargetMachine::TargetPassConfigCallbacks->clear();
+}
```
Interesting, yes for a codegen plugin this is actually a good end-to-end test!

weliveindetail replied:
Thanks! There will be a lot of LLVM configurations where this will likely fail (Windows, AIX, etc.) and we must figure out where it works and where not. Usually we make a good guess and go through a number of revert/reland cycles. It will be a pain though to do this with the whole patch and it would add unnecessary fuzz to the history for a lot of code. So, why not restrict it to the configuration that you are developing on and iterate from there? Iterations will only touch the test then.

@Tcc100 I guess this will be a good candidate. Is the test running for you like that?
```
# REQUIRES: native && system-linux && llvm-dylib
```

weliveindetail replied:
FYI: test passes for me on Ubuntu 22.04 with this config
```
> cmake -Sllvm -Bbuild -GNinja -DLLVM_BUILD_LLVM_DYLIB=On -DLLVM_LINK_LLVM_DYLIB=On ...
> ninja -C build CGPluginTest FileCheck llvm-config llc count not
> cd build && bin/llvm-lit -a --filter=codegen-plugin test
```

The `LLVM_DYLIB` options enable the LLVM shared library. This simplifies compatibility, because we always link the same ABI, even if we load the plugin e.g. in clang. It correctly shows up as unsupported without these options.

weliveindetail replied:
On Windows, the test plugin doesn't get built and the unittest is skipped as expected. The LIT test fails right now with:
```
Error opening 'S:\LLVM\build\bin/../unittests/CodeGen/CGPluginTest/CGTestPlugin.dll': Can't open: The specified module could not be found. (0x7E)
  -load request ignored.
```

Adding the proposed `REQUIRES` clause, it shows up as unsupported (BTW is should be commented as `; REQUIRES` of course). Let me know if you have questions about it.

Tcc100 replied:
Thanks for the detailed tests! I added `REQUIRES` (using commas which seems to be more common in the code).

weliveindetail commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -119,6 +125,9 @@ class TargetMachine {
 
 public:
+  static ManagedStatic<SmallVector<PassConfigCallback, 1>>
+      TargetPassConfigCallbacks;
```
Well, sounds reasonable, but I am still skeptic about the current implementation. I don't see `ManagedStatic` members LLVM header files anywhere.

Why not do this?
* add another static function like `invokeTargetPassConfigCallbacks([...])`
* move both declarations somewhere closer to `parseBinutilsVersion()` (which is static too)
* implement both functions in `TargetMachine.cpp` and define the vector there as well in global scope

Tcc100 replied:
I didn't use `ManagedStatic` to move the vector within the class but to solve the lifetime issue. The callbacks from the plugin must be unregistered before the shared object is unloaded. `ManagedStatic` was the straightforward way to solve this as `ManagedStatic`s are cleared right after LLVM finished.

I now adopted the cleaner way of other LLVM registries that bind the lifetime of the registration to the plugin using an object created in the plugin. This introduces the `RegisterTargetPassConfigCallback` class that automatically unregisters the callback when the plugin is unloaded.
I also created an invoke method and updated the examples accordingly. Please let me know if this fits your intent.

weliveindetail commented:
@Tcc100 Thanks for working on this! Once [this](https://github.com/llvm/llvm-project/pull/139059#discussion_r2175620644) and [this](https://github.com/llvm/llvm-project/pull/139059#discussion_r2175604034) got addressed, your patch should be ready to land.

Anyone other thoughts?

weliveindetail commented on llvm/unittests/CodeGen/CGPluginTest/PluginTest.cpp:
```diff
@@ -0,0 +1,114 @@
+  const auto &Argvs = testing::internal::GetArgvs();
+  if (Argvs.size() <= 0)
+    return std::nullopt;
+  void *Ptr = (void *)(intptr_t)anchor;
```
@Tcc100 Please prepare a PR for that when you find the time

weliveindetail commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -531,6 +532,39 @@ class LLVM_ABI TargetMachine {
+};
+
+class RegisterTargetPassConfigCallback
+    : public TargetPassConfigCallbackRegistry {
```
This is just a RAII wrapper right? Let's make the callback a member and not inherit from std::function

weliveindetail commented on llvm/include/llvm/Target/TargetMachine.h:
```diff
@@ -531,6 +532,39 @@ class LLVM_ABI TargetMachine {
+                              TargetPassConfig *PassConfig) {
+    for (const PassConfigCallback *C : Callbacks)
+      (*C)(TM, PM, PassConfig);
+  }
```
Do we need a registry? IIUC this could a simple freestanding function that is declared here and defined in the cpp (where you have the static member init right now). Maybe let the name express it, like `invokeGlobalTargetPassConfigCallbacks(...)`? We don't need the callbacks vector here in the header. It should be a global `ManagedStatic` then (instead of the static member init).

Tcc100 commented:
@weliveindetail I changed the registry as you suggested. Since it is now completely separate from the TargetMachine class, I moved it into its own file.

weliveindetail reviewed (approved):
Great, LGTM. @arsenm What do you think?

Let's keep the PR open for a few more days so everyone has the chance to add their notes. Otherwise I will merge it towards the end of the week.

nikic commented on llvm/lib/Target/RegisterTargetPassConfigCallback.cpp:
```diff
@@ -0,0 +1,26 @@
+#include "llvm/Target/RegisterTargetPassConfigCallback.h"
```
Missing license header.

weliveindetail replied:
Thanks, good catch! I will add it before landing

Tcc100 committed:
[CodeGen] Expose the extensibility of PassConfig to plugins

jplehr commented:
This broke https://lab.llvm.org/buildbot/#/builders/10/builds/9125/steps/8/logs/stdio

The bot does a SHARED_LIBS=ON build. I can reproduce locally with the CMake cache file in `offload/cmake/caches/AMDGPUBot.cmake` as the build config.
